solana-programs/anchor/audius-data/programs/audius-data/src/lib.rs

```diff
@@ -103,6 +109,175 @@ pub mod audius_data {
         Ok(())
     }
 
+    /// Create a content node account from the admin account.
+    /// The content node's account is derived from the admin base PDA + sp_id bytes.
+    pub fn create_content_node(
+        ctx: Context<CreateContentNode>,
+        base: Pubkey,
+        sp_id: u16,
+        authority: Pubkey,
+        owner_eth_address: [u8; 20],
+    ) -> Result<()> {
+        msg!("Audius::CreateContentNode");
+
+        // Confirm that the base used for content node account seed is derived from this Audius admin storage account
+        let (derived_base, _) = Pubkey::find_program_address(
+            &[&ctx.accounts.admin.key().to_bytes()[..32]],
+            ctx.program_id,
+        );
+
+        if derived_base != base {
+            return Err(ErrorCode::Unauthorized.into());
+        }
+
+        // Confirm that the derived pda from base is the same as the user storage account
+        let (derived_content_node, _) = Pubkey::find_program_address(
+            &[&derived_base.to_bytes()[..32], CONTENT_NODE_SEED_PREFIX, sp_id.to_le_bytes().as_ref()],
+            ctx.program_id,
+        );
+        if derived_content_node != ctx.accounts.content_node.key() {
+            return Err(ErrorCode::Unauthorized.into());
+        }
+
+        if ctx.accounts.authority.key() != ctx.accounts.admin.authority {
+            return Err(ErrorCode::Unauthorized.into());
+        }
+
+        let content_node = &mut ctx.accounts.content_node;
+        content_node.owner_eth_address = owner_eth_address;
+        content_node.authority = authority;
+
+        Ok(())
+    }
+
+    /// Create a content node account from other content nodes.
+    pub fn public_create_content_node(
+        ctx: Context<PublicCreateContentNode>,
+        base: Pubkey,
+        _p1: ProposerSeedBump,
+        _p2: ProposerSeedBump,
+        _p3: ProposerSeedBump,
+        sp_id: u16,
+        authority: Pubkey,
+        owner_eth_address: [u8; 20],
+    ) -> Result<()> {
+        msg!("Audius::PublicCreateContentNode");
+
+
+        // validate that there are no dupes in the replica
+        let proposers = [
+            (&ctx.accounts.proposer1, &ctx.accounts.proposer1_authority),
+            (&ctx.accounts.proposer2, &ctx.accounts.proposer2_authority),
+            (&ctx.accounts.proposer3, &ctx.accounts.proposer3_authority),
+        ];
+
+        // Ensure that no proposer's owner eth address is repeated
+        let mut eth_addresses = BTreeMap::new();
+        if !proposers.iter().all(move |(proposer, _)| return match eth_addresses.insert(proposer.owner_eth_address, true) {
+            Some(_) => false,
+            None => true
+        }) {
+            // duplicate owner eth address - err
+            return Err(ErrorCode::Unauthorized.into());
+        }
+
+        if !proposers.iter().all(|(proposer, authority)| proposer.authority == authority.key()) {
```

Comment:
is there a way to limit this dupe iteration somehow? as in we go through proposers 2x

Reply:
yeah, can change, good callout